### ISSUE TITLE
Travis CI build failing - zeromq context singleton

### ISSUE THREAD

**Issue by breml** [logstash-plugins/logstash-input-zeromq#22]
Currently, the travis ci build is failing: https://travis-ci.org/logstash-plugins/logstash-input-zeromq/builds/127623701

I reproduced the problem on my system. I think, the reason for this is the zeromq context singleton, introduced in https://github.com/logstash-plugins/logstash-mixin-zeromq/commit/53130a3e13a20ae1a67c82ff143b6142c0be9c5e

The problem is, in one spec test, the plugin is closed, which causes the zmq context to be terminated. Because `@context` in the ZeroMQContext-singleton (https://github.com/logstash-plugins/logstash-mixin-zeromq/blob/master/lib/logstash/plugin_mixins/zeromq.rb#L5) is not reset to `nil`, no new zmq context is created for the second spec test, which follows. This causes to failure to create a new zmq socket and finally leads to the error of calling `close` on a NilClass.

One way to make the spec-tests working again is to remove line https://github.com/logstash-plugins/logstash-input-zeromq/blob/master/lib/logstash/inputs/zeromq.rb#L136, where the ZMQ context is terminated.

In my opinion, there are two issues:
1. terminating the context is potentially called multiple times (multiple worker or even from different plugins like zmq input, filter and output).
2. terminating the context alone is not enough, the ZeroMQContext-singleton should also set `@context = nil` after the context is terminated.

I suggest the following solution:
- The zmq plugins (input, filter, output) should call `terminate_context` in their `close` method.
- The `terminate_context` method should check, if the context is not yet terminated. If so, the context should be terminated and `@context` should be set to `nil`.
- The `terminate_context` should be thread safe.

@avishai-ish-shalom

**Comment by breml** [comment]
Related to https://github.com/logstash-plugins/logstash-mixin-zeromq/issues/3

**Comment by nukemberg** [comment]
The underlying problem is that Logstash doesn't have global termination hooks. ZeroMQ termination needs to be orchestrated - as terminating the context releases blocked sockets. Having plugins call `terminate_context` is a poor man's workaround, but we may have to live with it.

**Comment by breml** [comment]
@avishai-ish-shalom I agree. For the moment, the poor man's workaround seems to be to only/best available solution.
